infoschema: refill more info schema cache to improve performance

wjhuang2016 opened the pull request:
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: close #64515

**Problem Summary:**

When executing `SHOW FULL TABLES` or similar operations that list tables in a database, TiDB with infoschema v2 always reads table metadata from TiKV, even when:
- The schema cache has sufficient capacity
- All tables in the database are already cached in memory

This causes unnecessary TiKV reads and slow query performance, especially for databases with thousands of tables.

### What changed and how does it work?

This PR optimizes the infoschema v2 cache behavior:

1. **Adjusted cache refill policy**: 
   - When cache usage is under 70% capacity, `TableByID` will automatically refill cache on miss (previously it never refilled by default)
   - This ensures we cache as many tables as possible while avoiding eviction of useful cached data

2. **Added fast path for `SchemaTableInfos`**:
   - When all tables in a database are already cached, read directly from cache instead of TiKV
   - Uses single-pass iteration to check cache and collect tables simultaneously (optimized per review feedback)
   - Falls back to TiKV read if any table is not cached

3. **New `tableFromCache()` helper method**:
   - Returns table directly from cache without triggering reload
   - Enables efficient cache lookup without side effects

**Performance improvement:**
- Before: `SHOW FULL TABLES` always reads from TiKV (~200ms+ for 3000 tables)
- After: Returns instantly from cache when all tables are cached

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
``` sql
Show full tables;
```
Before this PR:
<img width="303" height="433" alt="image" src="https://github.com/user-attachments/assets/4ac4c271-af18-451d-b167-e9a221077734" />


After this PR:
<img width="316" height="350" alt="image" src="https://github.com/user-attachments/assets/b136ed11-fdd8-4bbc-89ea-ab78f493b9f6" />


- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > <!-- Or your custom  "No need to test" reasons -->

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

wjhuang2016 committed:
init

Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

wjhuang2016 committed:
refine

Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

wjhuang2016 committed:
refine

Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

wjhuang2016 committed:
refine

Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

wjhuang2016 committed:
done

Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

wjhuang2016 committed:
refine comment

Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

joechenrh commented on pkg/infoschema/infoschema_v2.go:
```diff
@@ -1097,6 +1105,34 @@ func (is *infoschemaV2) SchemaTableInfos(ctx context.Context, schema ast.CIStr)
 	is.keepAlive()
+
+	allInCache := true
+	is.IterateAllTableItems(func(t TableItem) bool {
```
`IterateAllTableItems` will iterate `byName`, while `TableIsCached` also need to iterate `byID` again for each tableID. Is it redundant?

wjhuang2016 committed:
refine

Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

wjhuang2016 committed:
refine

Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

wjhuang2016 committed:
merge

Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

GMHDBJD commented on pkg/infoschema/infoschema_v2.go:
```diff
@@ -1097,6 +1117,42 @@ func (is *infoschemaV2) SchemaTableInfos(ctx context.Context, schema ast.CIStr)
+				tbl, ok := is.TableByID(ctx, t.TableID)
+				intest.Assert(ok, "invalid table id", t.DBName.L)
+				tables = append(tables, tbl.Meta())
+				return true
```
can we get the table via TableIsCached api? then we no need to get it again later if all in cache.

GMHDBJD reviewed (commented):
I remember that mysql client will exec `show full tables` when exec `use database`, please confirm that this meets our expectations for v2.

wjhuang2016 committed:
infoschema: address review comments

1. TestWithRefillOption: restore SchemaCacheSize after test
2. TableByID: update comment wording per suggestion
3. SchemaTableInfos: optimize fast path to avoid redundant lookups
   - Add tableFromCache() to get table directly from cache
   - Merge two iterations into one pass
   - Fix potential assertion failure when schema not exists

wjhuang2016 committed:
infoschema: fix fast path for SchemaTableInfos when no tables in btree

Add foundAny check to ensure we only use the fast path when at least
one table is found in the btree index. If no tables are found,
fallback to TiKV to handle the case where tables haven't been loaded
into the btree index yet.

wjhuang2016 committed:
Revert "infoschema: fix fast path for SchemaTableInfos when no tables in btree"

This reverts commit 85eb8ac6c7eb0fc4efd15435cd72678263cc9096.

wjhuang2016 committed:
Merge branch 'master' of https://github.com/pingcap/tidb into schema_cache_more

wjhuang2016 committed:
infoschema: fix iterateAllTableItemsInternal to use byName btree

Fix a bug where newly created tables with the same name in different
schemas were not visible in information_schema.tables.

The issue was that iterateAllTableItemsInternal used byID btree which
is sorted by {tableID, schemaVersion}, but the IterateAllTableItemsByDB
function passed a tableItem with dbID set. Since compareByID ignores
dbID, the iteration didn't work as expected.

Changed to use byName btree which is sorted by {dbName, tableName,
schemaVersion}, making the iteration work correctly.

Added regression test TestSameTableNameInDifferentSchemas.

wjhuang2016 committed:
executor: update BUILD.bazel for new test

wjhuang2016 commented:
/retest

wjhuang2016 committed:
infoschema: fix IterateAllTableItemsByDB search logic and SchemaTableInfos order
